qiskit_experiments/data_processing/data_processor.py

```diff
@@ -0,0 +1,118 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Actions done on the data to bring it in a usable form."""
+
+from typing import Any, Dict, List, Set, Tuple
+
+from qiskit_experiments.data_processing.data_action import DataAction
+from qiskit_experiments.data_processing.exceptions import DataProcessorError
+
+
+class DataProcessor:
+    """
+    A DataProcessor defines a sequence of operations to perform on experimental data.
+    Calling an instance of DataProcessor applies this sequence on the input argument.
+    A DataProcessor is created with a list of DataAction instances. Each DataAction
+    applies its _process method on the data and returns the processed data. The nodes
+    in the DataProcessor may also perform data validation and some minor formatting.
+    The output of one data action serves as input for the next data action.
+    DataProcessor.__call__(datum) usually takes in an entry from the data property of
+    an ExperimentData object (i.e. a dict containing metadata and memory keys and
+    possibly counts, like the Result.data property) and produces the formatted data.
+    DataProcessor.__call__(datum) extracts the data from the given datum under
+    DataProcessor._input_key (which is specified at initialization) of the given datum.
+    """
+
```

Comment:
This is actually a one-function class. As such, it is not clear if we really need a class here, or just two functions (of processing with or without history), that take as parameters `input_key`, `data_actions`, and `datum`. If we stick with a class, then need to think for each of these three parameters whether it belongs to the constructor or as a parameter to the processing function(s). My feeling is that `input_key` can move to the function.

Reply:
This might go against the idea of having the data_processor callable. @chriseclectic Thoughts on this?

Reply:
```
    def __call__(self, datum: Dict[str, Any]) -> Tuple[Any, Any]:
        """
        Call self on the given datum. This method sequentially calls the stored data actions
        on the datum.
        Args:
            datum: A single item of data, typically from an ExperimentData instance, that needs
                to be processed. This dict also contains the metadata of each experiment.
        Returns:
            processed data: The data processed by the data processor.
        Raises:
            DataProcessorError: if no nodes are present.
        """
        return self._call_internal(datum, False)

    def call_with_history(
        self, datum: Dict[str, Any], history_nodes: Set = None
    ) -> Tuple[Dict[str, Any], List]:
        """
        Call self on the given datum. This method sequentially calls the stored data actions
        on the datum and also returns the history of the processed data.
        Args:
            datum: A single item of data, typically from an ExperimentData instance, that
                needs to be processed.
            history_nodes: The nodes, specified by index in the data processing chain, to
                include in the history. If None is given then all nodes will be included
                in the history.
        Returns:
            processed data: The datum processed by the data processor.
            history: The datum processed at each node of the data processor.
        Raises:
            DataProcessorError: If no nodes are present.
        """
        return self._call_internal(datum, True, history_nodes)
       
    def _call_interna(self, datum, with_history (bool), history_nodes=None):
         if len(self._nodes) == 0:
            raise DataProcessorError("Cannot call an empty data processor.")

        if self._input_key not in datum:
            raise DataProcessorError(
                f"The input key {self._input_key} was not found in the input datum."
            )

        datum_ = datum[self._input_key]

        history = []
        for index, node in enumerate(self._nodes):
            datum_ = node(datum_)

            if with_history and (history_nodes is None or (history_nodes and index in history_nodes)):
                history.append((node.__class__.__name__, datum_, index))

        if with_history:
            return datum_, history
        else:
            return datum_
```

Reply:
Sorry, I posted the last comment inside the wrong conversation

Reply:
The point of it being a class is that all configuration is done via the class interface (for now in its init), and the final object can be treated like a function with a constant call signature, regardless of the type of data processing being performed.  If we just used the functions we would have to supply additional options for nodes etc that need to also be passed around everywhere this function would be used. Also later it would be good to add methods to the class to make it easier to modify/combine data processors (like adding nodes, conditionally dealing with multiple data types, chaining data processors together etc)